libupnp 1.10 support

```/var/tmp/portage/net-misc/gerbera-1.3.4/work/gerbera-1.3.4/src/server.cc: In member function ‘int Server::registerVirtualDirCallbacks()’:
/var/tmp/portage/net-misc/gerbera-1.3.4/work/gerbera-1.3.4/src/server.cc:468:20: error: invalid user-defined conversion from ‘Server::registerVirtualDirCallbacks()::<lambda(const char*, UpnpFileInfo*, const void*)>’ to ‘VDCallback_GetInfo’ {aka ‘int (*)(const char*, s_UpnpFileInfo*, const void*, const void**)’} [-fpermissive]
  468 |         return 0; });
      |                    ^
/var/tmp/portage/net-misc/gerbera-1.3.4/work/gerbera-1.3.4/src/server.cc:455:50: note: candidate is: ‘constexpr Server::registerVirtualDirCallbacks()::<lambda(const char*, UpnpFileInfo*, const void*)>::operator int (*)(const char*, UpnpFileInfo*, const void*)() const’ <near match>
```

See: https://bugs.gentoo.org/700394

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -355,6 +355,10 @@ if (UPNP_VERSION_STRING VERSION_GREATER_EQUAL "1.8.5")
     add_definitions(-DUPNP_HAS_EXTRA_HEADERS_LIST)
 endif()
 
+if (UPNP_VERSION_STRING VERSION_GREATER_EQUAL "1.10.0")
+    add_definitions(-DUPNP_HAS_REQUEST_COOKIES)
+endif()
+
 if (NOT UPNP_HAS_IPV6)
     message(FATAL_ERROR "Gerbera requires libupnp with IPv6 support.")
 endif()

--- a/src/server.cc
+++ b/src/server.cc
@@ -452,7 +452,11 @@ Ref<RequestHandler> Server::createRequestHandler(const char* filename) const
 int Server::registerVirtualDirCallbacks()
 {
     log_debug("Setting UpnpVirtualDir GetInfoCallback\n");
+#ifdef UPNP_HAS_REQUEST_COOKIES
+    int ret = UpnpVirtualDir_set_GetInfoCallback([](IN const char* filename, OUT UpnpFileInfo* info, const void* cookie, OUT const void** requestCookie) -> int {
+#else
     int ret = UpnpVirtualDir_set_GetInfoCallback([](IN const char* filename, OUT UpnpFileInfo* info, const void* cookie) -> int {
+#endif
         try {
             Ref<RequestHandler> reqHandler = static_cast<const Server *>(cookie)->createRequestHandler(filename);
             reqHandler->getInfo(filename, info);
@@ -470,7 +474,11 @@ int Server::registerVirtualDirCallbacks()
         return ret;
 
     log_debug("Setting UpnpVirtualDir OpenCallback\n");
+#ifdef UPNP_HAS_REQUEST_COOKIES
+    ret = UpnpVirtualDir_set_OpenCallback([](IN const char* filename, IN enum UpnpOpenFileMode mode, IN const void* cookie, IN const void* requestCookie) -> UpnpWebFileHandle {
+#else
     ret = UpnpVirtualDir_set_OpenCallback([](IN const char* filename, IN enum UpnpOpenFileMode mode, IN const void* cookie) -> UpnpWebFileHandle {
+#endif
         String link = urlUnescape(zmm::String::copy(filename));
 
         try {
@@ -493,7 +501,11 @@ int Server::registerVirtualDirCallbacks()
         return ret;
 
     log_debug("Setting UpnpVirtualDir ReadCallback\n");
+#ifdef UPNP_HAS_REQUEST_COOKIES
+    ret = UpnpVirtualDir_set_ReadCallback([](IN UpnpWebFileHandle f, OUT char* buf, IN size_t length, IN const void* cookie, IN const void* requestCookie) -> int {
+#else
     ret = UpnpVirtualDir_set_ReadCallback([](IN UpnpWebFileHandle f, OUT char* buf, IN size_t length, IN const void* cookie) -> int {
+#endif
         //log_debug("%p read(%d)\n", f, length);
         if (static_cast<const Server*>(cookie)->getShutdownStatus())
             return -1;
@@ -505,7 +517,11 @@ int Server::registerVirtualDirCallbacks()
         return ret;
 
     log_debug("Setting UpnpVirtualDir WriteCallback\n");
+#ifdef UPNP_HAS_REQUEST_COOKIES
+    ret = UpnpVirtualDir_set_WriteCallback([](IN UpnpWebFileHandle f, IN char* buf, IN size_t length, IN const void* cookie, IN const void* requestCookie) -> int {
+#else
     ret = UpnpVirtualDir_set_WriteCallback([](IN UpnpWebFileHandle f, IN char* buf, IN size_t length, IN const void* cookie) -> int {
+#endif
         //log_debug("%p write(%d)\n", f, length);
         return 0;
     });
@@ -513,7 +529,11 @@ int Server::registerVirtualDirCallbacks()
         return ret;
 
     log_debug("Setting UpnpVirtualDir SeekCallback\n");
+#ifdef UPNP_HAS_REQUEST_COOKIES
+    ret = UpnpVirtualDir_set_SeekCallback([](IN UpnpWebFileHandle f, IN off_t offset, IN int whence, IN const void* cookie, IN const void* requestCookie) -> int {
+#else
     ret = UpnpVirtualDir_set_SeekCallback([](IN UpnpWebFileHandle f, IN off_t offset, IN int whence, IN const void* cookie) -> int {
+#endif
         //log_debug("%p seek(%d, %d)\n", f, offset, whence);
         try {
             auto* handler = static_cast<IOHandler*>(f);
@@ -530,7 +550,11 @@ int Server::registerVirtualDirCallbacks()
         return ret;
 
     log_debug("Setting UpnpVirtualDir CloseCallback\n");
+#ifdef UPNP_HAS_REQUEST_COOKIES
+    UpnpVirtualDir_set_CloseCallback([](IN UpnpWebFileHandle f, IN const void* cookie, IN const void* requestCookie) -> int {
+#else
     UpnpVirtualDir_set_CloseCallback([](IN UpnpWebFileHandle f, IN const void* cookie) -> int {
+#endif
         //log_debug("%p close()\n", f);
         Ref<IOHandler> handler((IOHandler*)f);
         handler->release();

